File: netDxf/Entities/PeriodicSplineExactEvaluation.cs
using System;
using BigInteger = System.Numerics.BigInteger;

namespace netDxf.Entities
{
    /// <summary>Bounded exact evaluation for ill-conditioned periodic samples.</summary>
    internal static class PeriodicSplineExactEvaluation
    {
        // A sample uses at most eleven controls and 55 local basis steps. The
        // explicit component budget also bounds pathological rational growth.
        private const int MaximumIntegerBytes = 65536;

        internal static Vector3 Evaluate(Vector3[] expandedControls, double[] expandedWeights,
            double[] knots, int degree, double parameter, int firstControl)
        {
            if (degree < 1 || degree > Spline.MaxDegree)
                throw new ArgumentException("Exact periodic SPLINE evaluation requires a supported degree.");
            Rational u = Rational.FromDouble(parameter);
            var localKnots = new Rational[2 * degree + 1];
            for (int i = 0; i < localKnots.Length; i++)
                localKnots[i] = Rational.FromDouble(knots[firstControl + i]);

            // Local nonzero basis recurrence (Algorithm A2.2), evaluated over
            // exact rational values of the actual binary64 inputs. Reusing a
            // rounded double basis would retain the cancellation defect.
            var basis = new Rational[degree + 1];
            var left = new Rational[degree + 1];
            var right = new Rational[degree + 1];
            basis[0] = Rational.One;
            for (int j = 1; j <= degree; j++)
            {
                left[j] = u - localKnots[degree + 1 - j];
                right[j] = localKnots[degree + j] - u;
                Rational saved = Rational.Zero;
                for (int r = 0; r < j; r++)
                {
                    Rational divisor = right[r + 1] + left[j - r];
                    if (divisor.Sign <= 0)
                        throw new ArgumentException("Exact periodic SPLINE evaluation requires increasing knots.");
                    Rational temporary = basis[r] / divisor;
                    basis[r] = saved + right[r + 1] * temporary;
                    saved = left[j - r] * temporary;
                }
                basis[j] = saved;
            }

            Rational denominator = Rational.Zero;
            Rational x = Rational.Zero, y = Rational.Zero, z = Rational.Zero;
            for (int i = 0; i <= degree; i++)
            {
                Rational weighted = basis[i] * Rational.FromDouble(expandedWeights[firstControl + i]);
                if (weighted.Sign < 0)
                    throw new ArgumentException("Exact periodic SPLINE evaluation requires positive rational weights.");
                if (weighted.Sign == 0) continue;
                Vector3 point = expandedControls[firstControl + i];
                denominator += weighted;
                x += weighted * Rational.FromDouble(point.X);
                y += weighted * Rational.FromDouble(point.Y);
                z += weighted * Rational.FromDouble(point.Z);
            }
            if (denominator.Sign <= 0)
                throw new ArgumentException("The exact periodic SPLINE denominator is zero.");
            return new Vector3((x / denominator).ToDouble(), (y / denominator).ToDouble(), (z / denominator).ToDouble());
        }

        private static void CheckSize(BigInteger value)
        {
            if (value.ToByteArray().Length > MaximumIntegerBytes)
                throw new NotSupportedException("Exact periodic SPLINE evaluation exceeds its bounded arithmetic budget.");
        }

        private static BigInteger Product(BigInteger a, BigInteger b)
        {
            if (a.IsZero || b.IsZero) return BigInteger.Zero;
            if ((long)a.ToByteArray().Length + b.ToByteArray().Length > MaximumIntegerBytes)
                throw new NotSupportedException("Exact periodic SPLINE evaluation exceeds its bounded arithmetic budget.");
            return a * b;
        }

        private readonly struct Rational
        {
            internal static readonly Rational Zero = new Rational(BigInteger.Zero, BigInteger.One, false);
            internal static readonly Rational One = new Rational(BigInteger.One, BigInteger.One, false);
            private readonly BigInteger numerator;
            private readonly BigInteger denominator;

            private Rational(BigInteger numerator, BigInteger denominator, bool reduce = true)
            {
                if (denominator.IsZero) throw new ArgumentException("An exact periodic SPLINE divisor is zero.");
                if (numerator.IsZero)
                { this.numerator = BigInteger.Zero; this.denominator = BigInteger.One; return; }
                if (denominator.Sign < 0) { numerator = -numerator; denominator = -denominator; }
                CheckSize(numerator); CheckSize(denominator);
                if (reduce)
                {
                    BigInteger divisor = BigInteger.GreatestCommonDivisor(BigInteger.Abs(numerator), denominator);
                    numerator /= divisor; denominator /= divisor;
                }
                this.numerator = numerator;
                this.denominator = denominator;
            }

            internal int Sign { get { return this.numerator.Sign; } }

            internal static Rational FromDouble(double value)
            {
                long bits = BitConverter.DoubleToInt64Bits(value);
                int field = (int)((bits >> 52) & 0x7ff);
                if (field == 0x7ff) throw new ArgumentException("Exact periodic SPLINE inputs must be finite.");
                long significand = bits & 0x000fffffffffffffL;
                if (field != 0) significand |= 1L << 52;
                if (significand == 0) return Zero;
                BigInteger integer = new BigInteger(significand);
                if (bits < 0) integer = -integer;
                int exponent = field == 0 ? -1074 : field - 1075;
                return exponent >= 0 ? new Rational(integer << exponent, BigInteger.One, false)
                    : new Rational(integer, BigInteger.One << -exponent);
            }

            public static Rational operator +(Rational a, Rational b)
            {
                if (a.numerator.IsZero) return b;
                if (b.numerator.IsZero) return a;
                BigInteger shared = BigInteger.GreatestCommonDivisor(a.denominator, b.denominator);
                BigInteger leftScale = b.denominator / shared, rightScale = a.denominator / shared;
                return new Rational(Product(a.numerator, leftScale) + Product(b.numerator, rightScale),
                    Product(a.denominator, leftScale));
            }

            public static Rational operator -(Rational a, Rational b)
            { return a + new Rational(-b.numerator, b.denominator, false); }

            public static Rational operator *(Rational a, Rational b)
            {
                if (a.numerator.IsZero || b.numerator.IsZero) return Zero;
                BigInteger first = BigInteger.GreatestCommonDivisor(BigInteger.Abs(a.numerator), b.denominator);
                BigInteger second = BigInteger.GreatestCommonDivisor(BigInteger.Abs(b.numerator), a.denominator);
                return new Rational(Product(a.numerator / first, b.numerator / second),
                    Product(a.denominator / second, b.denominator / first), false);
            }

            public static Rational operator /(Rational a, Rational b)
            {
                if (b.numerator.IsZero) throw new ArgumentException("An exact periodic SPLINE divisor is zero.");
                return a * new Rational(b.denominator, b.numerator, false);
            }

            internal double ToDouble()
            {
                if (this.numerator.IsZero) return 0.0;
                long sign = this.numerator.Sign < 0 ? long.MinValue : 0;
                BigInteger magnitude = BigInteger.Abs(this.numerator);
                int exponent = BitLength(magnitude) - BitLength(this.denominator);
                if (exponent >= 0)
                {
                    if (magnitude < (this.denominator << exponent)) exponent--;
                }
                else if ((magnitude << -exponent) < this.denominator) exponent--;
                if (exponent > 1023)
                    throw new ArgumentException("The exact periodic SPLINE sample exceeds the finite binary64 range.");

                // Divide once at the destination's spacing. This avoids the
                // double rounding of separately rounded numerator/denominator
                // or a normal result subsequently scaled into subnormal range.
                int spacing = exponent < -1022 ? -1074 : exponent - 52;
                BigInteger dividend = magnitude, divisor = this.denominator;
                if (spacing < 0) dividend <<= -spacing;
                else divisor <<= spacing;
                BigInteger remainder;
                BigInteger significand = BigInteger.DivRem(dividend, divisor, out remainder);
                int comparison = (remainder << 1).CompareTo(divisor);
                if (comparison > 0 || comparison == 0 && !significand.IsEven) significand++;

                if (exponent < -1022)
                    return BitConverter.Int64BitsToDouble(sign | (long)significand);
                if (significand == (BigInteger.One << 53))
                { significand >>= 1; exponent++; }
                if (exponent > 1023)
                    throw new ArgumentException("The exact periodic SPLINE sample rounds outside the finite binary64 range.");
                long fraction = (long)(significand - (BigInteger.One << 52));
                return BitConverter.Int64BitsToDouble(sign | ((long)(exponent + 1023) << 52) | fraction);
            }

            private static int BitLength(BigInteger positive)
            {
                byte[] bytes = positive.ToByteArray();
                int last = bytes.Length - 1;
                while (last > 0 && bytes[last] == 0) last--;
                int high = bytes[last], bits = 0;
                while (high != 0) { bits++; high >>= 1; }
                return last * 8 + bits;
            }
        }
    }
}

File: javascript/netDxf/Entities/PeriodicSplineExactEvaluation.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { Vector3 } from '../Vector3.js';
import { ArgumentException, NotSupportedException } from '../../runtime/Errors.js';
const bits=new DataView(new ArrayBuffer(8));
const abs=x=>x<0n?-x:x;
const length=x=>x===0n?0:x.toString(2).length;
const bytes=x=>Math.max(1,Math.ceil((length(x<0n?-x-1n:x)+1)/8));
function check(x){if(bytes(x)>65536)throw new NotSupportedException('Exact periodic arithmetic budget exceeded.');}
function product(a,b){if(a===0n||b===0n)return 0n;if(bytes(a)+bytes(b)>65536)throw new NotSupportedException('Exact periodic arithmetic budget exceeded.');return a*b;}
function gcd(a,b){while(b!==0n){const c=a%b;a=b;b=c;}return a;}
class Rational {
  constructor(n,d=1n,reduce=true){
    if(d===0n)throw new ArgumentException('An exact periodic divisor is zero.');
    if(n===0n){this.n=0n;this.d=1n;return;}if(d<0n){n=-n;d=-d;}check(n);check(d);
    if(reduce){const g=gcd(abs(n),d);n/=g;d/=g;}this.n=n;this.d=d;
  }
  static FromDouble(v){
    bits.setFloat64(0,v);const raw=bits.getBigUint64(0),field=Number((raw>>52n)&2047n);
    if(field===2047)throw new ArgumentException('Exact periodic inputs must be finite.');
    let n=raw&0xfffffffffffffn;if(field!==0)n|=1n<<52n;if(n===0n)return zero;
    if(raw>>63n)n=-n;const e=field===0?-1074:field-1075;
    return e>=0?new Rational(n<<BigInt(e),1n,false):new Rational(n,1n<<BigInt(-e));
  }
  add(b){if(this.n===0n)return b;if(b.n===0n)return this;const shared=gcd(this.d,b.d),ls=b.d/shared,rs=this.d/shared;return new Rational(product(this.n,ls)+product(b.n,rs),product(this.d,ls));}
  sub(b){return this.add(new Rational(-b.n,b.d,false));}
  mul(b){if(this.n===0n||b.n===0n)return zero;const a=gcd(abs(this.n),b.d),c=gcd(abs(b.n),this.d);return new Rational(product(this.n/a,b.n/c),product(this.d/c,b.d/a),false);}
  div(b){if(b.n===0n)throw new ArgumentException('An exact periodic divisor is zero.');return this.mul(new Rational(b.d,b.n,false));}
  ToDouble(){
    if(this.n===0n)return 0;const sign=this.n<0n?1n<<63n:0n,m=abs(this.n);let e=length(m)-length(this.d);
    if(e>=0){if(m<(this.d<<BigInt(e)))e--;}else if((m<<BigInt(-e))<this.d)e--;
    if(e>1023)throw new ArgumentException('Exact periodic sample exceeds binary64 range.');
    const spacing=e<-1022?-1074:e-52;let dividend=m,divisor=this.d;
    if(spacing<0)dividend<<=BigInt(-spacing);else divisor<<=BigInt(spacing);
    let s=dividend/divisor;const rem=(dividend%divisor)<<1n;if(rem>divisor||(rem===divisor&&(s&1n)!==0n))s++;
    if(e<-1022){bits.setBigUint64(0,sign|s);return bits.getFloat64(0);}
    if(s===1n<<53n){s>>=1n;e++;}if(e>1023)throw new ArgumentException('Exact periodic sample rounds outside binary64 range.');
    bits.setBigUint64(0,sign|(BigInt(e+1023)<<52n)|(s-(1n<<52n)));return bits.getFloat64(0);
  }
}
const zero=new Rational(0n,1n,false),one=new Rational(1n,1n,false);
export class PeriodicSplineExactEvaluation {
  static Evaluate(controls,weights,knots,degree,parameter,first){
    if(degree<1||degree>10)throw new ArgumentException('Exact periodic evaluation requires a supported degree.');
    const u=Rational.FromDouble(parameter),local=Array.from({length:2*degree+1},(_,i)=>Rational.FromDouble(knots[first+i]));
    const basis=Array(degree+1).fill(zero),left=[],right=[];basis[0]=one;
    for(let j=1;j<=degree;j++){
      left[j]=u.sub(local[degree+1-j]);right[j]=local[degree+j].sub(u);let saved=zero;
      for(let k=0;k<j;k++){
        const divisor=right[k+1].add(left[j-k]);if(divisor.n<=0n)throw new ArgumentException('Exact periodic knots must increase.');
        const t=basis[k].div(divisor);basis[k]=saved.add(right[k+1].mul(t));saved=left[j-k].mul(t);
      }basis[j]=saved;
    }
    let den=zero,x=zero,y=zero,z=zero;
    for(let i=0;i<=degree;i++){
      const w=basis[i].mul(Rational.FromDouble(weights[first+i]));if(w.n<0n)throw new ArgumentException('Exact periodic weights must be positive.');if(w.n===0n)continue;
      const p=controls[first+i];den=den.add(w);x=x.add(w.mul(Rational.FromDouble(p.X)));y=y.add(w.mul(Rational.FromDouble(p.Y)));z=z.add(w.mul(Rational.FromDouble(p.Z)));
    }
    if(den.n<=0n)throw new ArgumentException('Exact periodic denominator is zero.');
    return new Vector3(x.div(den).ToDouble(),y.div(den).ToDouble(),z.div(den).ToDouble());
  }
}
